### PR TITLE
Add IntervalsSource for range and regexp queries

### PR DESCRIPTION
**Pull request by mayya-sharipova**
We already have convenient functions for contructing IntervalsSource for wildcard and fuzzy functions. This adds functions for regexp and range as well.

### CODE REVIEW

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java [hunk 1]
```diff
@@ -206,6 +210,91 @@ public static IntervalsSource wildcard(BytesRef wildcard, int maxExpansions) {
     return new MultiTermIntervalsSource(ca, maxExpansions, wildcard.utf8ToString());
   }
 
+  /**
+   * Return an {@link IntervalsSource} over the disjunction of all terms that match a regular
+   * expression
+   *
+   * <p>WARNING: Setting {@code maxExpansions} to higher than the default value of {@link
```
Comment (romseygeek) [inline lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java:217]: This line doesn't need to be on this method

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java [hunk 2]
```diff
@@ -206,6 +210,91 @@ public static IntervalsSource wildcard(BytesRef wildcard, int maxExpansions) {
...
+   * #DEFAULT_MAX_EXPANSIONS} can be both slow and memory-intensive
+   *
+   * @param regexp regula expression
```
Comment (romseygeek) [inline lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java:220]:
```suggestion
   * @param regexp regular expression
```

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java [hunk 3]
```diff
@@ -206,6 +210,91 @@ public static IntervalsSource wildcard(BytesRef wildcard, int maxExpansions) {
...
+  }
+
+  /**
+   * Expert: Return an {@link IntervalsSource} over the disjunction of all terms that match a
+   * regular expression
+   *
+   * <p>WARNING: Setting {@code maxExpansions} to higher than the default value of {@link
+   * #DEFAULT_MAX_EXPANSIONS} can be both slow and memory-intensive
+   *
+   * @param regexp regula expression
```
Comment (romseygeek) [inline lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java:236]:
```suggestion
   * @param regexp regular expression
```

**Review by romseygeek — APPROVED** [review]
A couple of javadoc nits, but LGTM otherwise

**Review by dweiss — APPROVED** [review]

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java (hunk 1 above)
Comment (mayya-sharipova) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java]: @romseygeek Thanks for feedback. Very nice to see you again. Addressed in  bdc43c7256da8dc178ec33d307d18d47de80ebeb

File: lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java (hunk 3 above)
Comment (mayya-sharipova) [reply lucene/queries/src/java/org/apache/lucene/queries/intervals/Intervals.java]: Addressed in bdc43c7256da8dc178ec33d307d18d47de80ebebbdc43c7256da8dc178ec33d307d18d47de80ebeb

**Comment by mayya-sharipova** [comment]
@romseygeek @dweiss Thank you for the review.

**Pull request merged**